frame/assets/src/lib.rs

```diff
@@ -569,45 +592,207 @@ pub mod pallet {
+				Ok(_) => None,
+				Err(origin) => Some(ensure_signed(origin)?),
+			};
+			let _ = Asset::<T, I>::try_mutate_exists(
+				id,
+				|maybe_details| -> Result<(), DispatchError> {
+					let mut details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+					if let Some(check_owner) = maybe_check_owner {
+						ensure!(details.owner == check_owner, Error::<T, I>::NoPermission);
+					}
+					ensure!(details.is_frozen, Error::<T, I>::BadWitness);
+					details.status = AssetStatus::Destroying;
+
+					Self::deposit_event(Event::DestructionStarted { asset_id: id });
+					Ok(())
+				},
+			)?;
+			Ok(())
+		}
+
+		/// Destroy all accounts associated with a given asset.
+		/// `destroy_accounts` should only be called after `start_destroy` has been called, and the
+		/// asset is in a `Destroying` state
+		///
+		/// Due to weight restrictions, this function may need to be called multiple
+		/// times to fully destroy all accounts. It will destroy `RemoveKeysLimit` accounts at a
+		/// time.
+		///
+		/// The origin must conform to `ForceOrigin` or must be Signed and the sender must be the
+		/// owner of the asset `id`.
+		///
+		/// - `id`: The identifier of the asset to be destroyed. This must identify an existing
+		///   asset.
+		///
+		/// Each call Emits the `Event::DestroyedAccounts` event.
+		#[pallet::weight(T::WeightInfo::destroy_accounts(T::RemoveKeysLimit::get()))]
+		pub fn destroy_accounts(
 			origin: OriginFor<T>,
 			#[pallet::compact] id: T::AssetId,
-			witness: DestroyWitness,
 		) -> DispatchResultWithPostInfo {
 			let maybe_check_owner = match T::ForceOrigin::try_origin(origin) {
 				Ok(_) => None,
 				Err(origin) => Some(ensure_signed(origin)?),
 			};
-			let details = Self::do_destroy(id, witness, maybe_check_owner)?;
-			Ok(Some(T::WeightInfo::destroy(
-				details.accounts.saturating_sub(details.sufficients),
-				details.sufficients,
-				details.approvals,
-			))
-			.into())
+			let mut dead_accounts: Vec<T::AccountId> = vec![];
+			let mut removed_accounts = 0;
+			let _ = Asset::<T, I>::try_mutate_exists(
+				id,
+				|maybe_details| -> Result<(), DispatchError> {
+					let mut details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+					if let Some(check_owner) = maybe_check_owner {
+						ensure!(details.owner == check_owner, Error::<T, I>::NoPermission);
+					}
+					ensure!(details.is_frozen, Error::<T, I>::BadWitness);
+					// Should only destroy accounts while the asset is being destroyed
+					ensure!(details.status == AssetStatus::Destroying, Error::<T, I>::LiveAsset);
+
+					for (who, v) in Account::<T, I>::drain_prefix(id) {
+						let _ = Self::dead_account(&who, &mut details, &v.reason, true);
+						dead_accounts.push(who);
+						removed_accounts = removed_accounts.saturating_add(1);
+						if removed_accounts >= T::RemoveKeysLimit::get() {
+							break
+						}
+					}
+
+					Self::deposit_event(Event::DestroyedAccounts {
```

Comment:
wonder if its valid to emit event from try mutate callback, cant it fail?

Reply:
I saw it done in the other extrinsics, so i just did the same. But I don't know enough about this.